File: Kern/Sources/DuoKern/Box/BoxEngine.swift
import Foundation

/// The growing-box engine: pure functions over `BoxState`.
///
/// Time discipline: every API takes `now`/`calendar` from the caller —
/// nothing in here ever calls `Date()` or `Calendar.current`.
/// All scheduling reads/writes are scoped to `state.config.direction`.
public enum BoxEngine {

    /// Fresh state from imported cards (nothing scheduled yet).
    public static func bootstrap(cards: [Card], config: BoxConfig) -> BoxState {
        var byID = [String: Card](minimumCapacity: cards.count)
        for card in cards {
            byID[card.id] = card
        }
        return BoxState(config: config, cards: byID)
    }

    /// Append ids to the user priority queue.
    /// Enqueuing a phrase auto-prioritizes its missing (unscheduled) component
    /// words ahead of it (design §Box 4). Already-scheduled or unknown ids are
    /// skipped; duplicates are dropped. Enqueued cards lead composition and
    /// bypass the health gate, but respect the load throttle (`maxLearning`):
    /// a pack enrolls and drips in at the pool rate, it is not dumped at once.
    public static func enqueue(state: BoxState, cardIDs: [String]) -> BoxState {
        var next = state
        var queued = Set(next.enqueued)

        func append(_ id: String) {
            guard next.cards[id] != nil,
                  scheduling(next, id) == nil,
                  !queued.contains(id) else { return }
            next.enqueued.append(id)
            queued.insert(id)
        }

        for id in cardIDs {
            for componentID in next.cards[id]?.componentIDs ?? [] {
                append(componentID)
            }
            append(id)
        }
        return next
    }

    /// Suspend or revive a card in the current direction.
    /// No-op for cards without scheduling (a card that was never introduced
    /// has nothing to suspend).
    public static func setSuspended(state: BoxState, cardID: String, suspended: Bool) -> BoxState {
        var next = state
        let key = BoxState.schedulingKey(cardID: cardID, direction: state.config.direction)
        guard var sched = next.scheduling[key] else { return next }
        sched.suspended = suspended
        next.scheduling[key] = sched
        return next
    }

    /// Session end: fold today's counters into `dailyStats` and prune
    /// `newIntroduced` to the trailing 60 days. Multiple sessions on one day
    /// accumulate `reviews`; `introduced`/`activeCount` reflect the latest state.
    public static func endSession(state: BoxState, reviewsDone: Int, now: Date, calendar: Calendar) -> BoxState {
        var next = state
        let day = dayKey(for: now, calendar: calendar)

        var stats = next.dailyStats[day] ?? DayStats()
        stats.reviews += reviewsDone
        stats.introduced = next.newIntroduced[day] ?? 0
        stats.activeCount = activeSchedulings(next).count
        next.dailyStats[day] = stats

        if let cutoff = calendar.date(byAdding: .day, value: -59, to: calendar.startOfDay(for: now)) {
            let cutoffKey = dayKey(for: cutoff, calendar: calendar)
            // why: yyyy-MM-dd keys compare chronologically as strings,
            // so pruning is a plain string comparison.
            next.newIntroduced = next.newIntroduced.filter { $0.key >= cutoffKey }
        }
        return next
    }
}

File: kmp/kern/src/commonMain/kotlin/net/spross/kern/box/BoxEngine.kt
package net.spross.kern.box

import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.minus
import net.spross.kern.model.BoxConfig
import net.spross.kern.model.Card
import net.spross.kern.model.DayStats
import net.spross.kern.model.JoinStamp
import net.spross.kern.model.Rating
import net.spross.kern.model.UnitKey

/**
 * The growing-box engine: pure functions over [BoxState].
 *
 * Time discipline: every API takes `nowEpochMillis`/`tzId` from the caller —
 * nothing in here ever reads a clock or the device calendar.
 * Denomination: growth is counted in CONCEPTS, workload (due/session caps) in UNITS.
 */
object BoxEngine {

    /** Fresh state for a (source, target) join; nothing scheduled yet. */
    fun bootstrap(cards: List<Card>, config: BoxConfig, joinStamp: JoinStamp): BoxState =
        BoxState(config = config, cards = cards.associateBy { it.id }, joinStamp = joinStamp)

    /**
     * Swap the join (source switch or catalog update) keeping every schedule, queue
     * entry, and stat: entries whose card no longer joins turn inert and revive here
     * on switch-back.
     */
    fun rejoin(state: BoxState, cards: List<Card>, joinStamp: JoinStamp): BoxState =
        state.copy(cards = cards.associateBy { it.id }, joinStamp = joinStamp)

    /**
     * Append CONCEPT ids to the user priority queue. Enqueuing a phrase auto-prepends
     * its missing (produce-unscheduled) components ahead of it. Unknown/non-joining ids,
     * already-scheduled concepts, and duplicates are skipped. Enqueued concepts lead
     * composition and bypass the health gate, but respect the load throttle: a pack
     * enrolls and drips in at the pool rate, it is not dumped at once.
     */
    fun enqueue(state: BoxState, conceptIds: List<String>): BoxState {
        val queued = state.enqueued.toMutableList()
        val seen = state.enqueued.toMutableSet()

        fun append(id: String) {
            if (id in seen || state.cards[id] == null) return
            if (state.scheduling[UnitKey.produce(id).encoded] != null) return
            queued += id
            seen += id
        }

        for (id in conceptIds) {
            state.cards[id]?.components?.forEach(::append)
            append(id)
        }
        return state.copy(enqueued = queued)
    }

    /** Suspend or revive ONE unit; no-op when the key has no schedule. */
    fun setSuspended(state: BoxState, unitKey: String, suspended: Boolean): BoxState {
        val sched = state.scheduling[unitKey] ?: return state
        return state.copy(
            scheduling = state.scheduling + (unitKey to sched.copy(suspended = suspended)),
        )
    }

    /**
     * Apply one answer to a unit. Introduction = the unit's first answer: creates its
     * schedule, and on a PRODUCE intro counts the concept introduced and dequeues it.
     * Introductions re-check eligibility and the concept budget at answer time
     * ([AnswerStatus.DroppedIneligible] / [AnswerStatus.DroppedPoolFull]) — plans
     * outlive phase changes. Review-phase Again answers count lapses; 8 lapses
     * auto-suspend the unit (leech). Non-joining or unknown keys are a defined no-op
     * ([AnswerStatus.StaleUnit]).
     */
    fun answer(
        state: BoxState,
        unitKey: String,
        rating: Rating,
        nowEpochMillis: Long,
        tzId: String,
    ): AnswerOutcome = Answering.answer(state, unitKey, rating, nowEpochMillis, tzId)

    /**
     * Fold the session into `dailyStats` (reviews accumulate in UNITS; introduced and
     * activeCount are CONCEPTS) and prune `newIntroduced` to the trailing 60 days.
     */
    fun endSession(state: BoxState, reviewsDone: Int, nowEpochMillis: Long, tzId: String): BoxState {
        val day = dayKey(nowEpochMillis, tzId)
        val previous = state.dailyStats[day] ?: DayStats()
        val folded = DayStats(
            reviews = previous.reviews + reviewsDone,
            introduced = state.newIntroduced[day] ?: 0,
            activeCount = Inventory.activeConceptCount(state),
        )
        // why: yyyy-MM-dd keys compare chronologically as strings, so pruning is a
        // plain string comparison.
        val cutoff = localDate(nowEpochMillis, tzId).minus(59, DateTimeUnit.DAY).toString()
        return state.copy(
            dailyStats = state.dailyStats + (day to folded),
            newIntroduced = state.newIntroduced.filterKeys { it >= cutoff },
        )
    }

    /** Joined, active unit keys due at `now`, oldest first — the drain-loop feed. */
    fun dueNow(state: BoxState, nowEpochMillis: Long): List<String> =
        Inventory.due(state, nowEpochMillis).map { it.key }

    fun statistics(state: BoxState, nowEpochMillis: Long, tzId: String): BoxStatistics =
        Statistics.statistics(state, nowEpochMillis, tzId)

    /** See [Exposure.exposureCards]; `nowEpochMillis` reserved for future due-weighting. */
    fun exposureCards(state: BoxState, nowEpochMillis: Long, limit: Int): List<Card> =
        Exposure.exposureCards(state, limit)
}
